react_juce/core/EcmascriptEngine_Hermes.cpp

```diff
@@ -0,0 +1,522 @@
+        }
+
+        juce::var::NativeFunction jsiMethodToVarMethod(jsi::Function v, jsi::Runtime &runtime)
+        {
+            auto fPtr = std::make_shared<CopyableJSIMethodWrapper>(std::move(v));
+
+            return [fPtr = fPtr, &rt = runtime] (const juce::var::NativeFunctionArgs &args)
+            {
+                std::vector<jsi::Value> jsiArgs;
+                for (int i = 0; i < args.numArguments; ++i)
+                {
+                    jsiArgs.emplace_back(varToJSIValue(args.arguments[i], rt));
+                }
+
+                const jsi::Value *jsiArgsPtr = jsiArgs.data();
+                return jsiValueToVar(fPtr->fn.call(rt, jsiArgsPtr, jsiArgs.size()), rt);
+            };
+        }
+
+        juce::var jsiObjectToVarObject(jsi::HostObject &v, jsi::Runtime &runtime)
+        {
+            juce::DynamicObject::Ptr varObj = new juce::DynamicObject();
+
+            auto props = v.getPropertyNames(runtime);
+            for (auto &propID : props)
+            {
+                jsi::Value property = v.get(runtime, propID);
+                varObj->setProperty(juce::String(propID.utf8(runtime)), jsiValueToVar(property, runtime));
+            }
+
+            return varObj.get();
+        }
+
+        juce::var jsiValueToVar(const jsi::Value &v, jsi::Runtime &runtime)
+        {
+            if (v.isBool())
+                return v.getBool();
+
+            if (v.isNumber())
+                return v.getNumber();
+
+            if (v.isString())
+                return juce::String(v.getString(runtime).utf8(runtime));
+
+            if (v.isUndefined())
+                return juce::var::undefined();
+
+            if (v.isNull())
+                return juce::var();
+
+            if (v.isObject())
+            {
+                auto obj = v.getObject(runtime);
+
+                if (obj.isArray(runtime))
+                    return jsiArrayToVarArray(obj.getArray(runtime), runtime);
+
+                if (obj.isFunction(runtime))
+                    return jsiMethodToVarMethod(obj.asFunction(runtime), runtime);
+
+                if (obj.isHostObject(runtime))
+                {
+                    auto hostObjectPtr = obj.getHostObject(runtime);
+                    return jsiObjectToVarObject(*hostObjectPtr, runtime);
+                }
+
+                if (obj.isArrayBuffer(runtime))
+                {
+                    jsi::ArrayBuffer arrayBuffer = obj.getArrayBuffer(runtime);
+                    juce::MemoryBlock memBlock(static_cast<const void*>(arrayBuffer.data(runtime)),
+                                                                        arrayBuffer.size(runtime));
+
+                    return juce::var(memBlock);
+                }
+            }
+
+            return {};
+        }
+
+        //==============================================================================
+        jsi::Object varObjectToJSIObject(juce::DynamicObject *v, jsi::Runtime &runtime)
```

Comment:
@nick-thompson , so this is where I have an implementation question. 

At the moment this conversion assumes that the `juce::DynamicObject` instance passed in will live to match the lifetime of the equivalent `jsi` object type. We make this same assumption in the duktape implementation. The same concern applies to the other side of the conversion `jsiObjectToVarObject`. Do we think this is fine? The other approach I had for this to ensure lifetime safety was as below:

```
+        //==============================================================================
+        class JSIDynamicObject : public jsi::HostObject
+        {
+        public:
+            explicit JSIDynamicObject(juce::DynamicObject *obj)
+                : objPtr(obj)
+            {
+                jassert(objPtr);
+            }
+
+            jsi::Value get(jsi::Runtime& rt, const jsi::PropNameID& name) override
+            {
+                return varToJSIValue(objPtr->getProperty(name.utf8(rt).c_str()), rt);
+            }
+
+            void set(jsi::Runtime& rt, const jsi::PropNameID& name, const jsi::Value& value)
+            {
+                objPtr->setProperty(name.utf8(rt).c_str(), jsiValueToVar(value, rt));
+            }
+
+            std::vector<jsi::PropNameID> getPropertyNames(jsi::Runtime& rt)
+            {
+                std::vector<jsi::PropNameID> props;
+                for (auto &p : objPtr->getProperties())
+                {
+                    props.push_back(jsi::PropNameID::forAscii(rt, p.name.getCharPointer()));
+                }
+
+                return props;
+            }
+
+        private:
+            juce::DynamicObject::Ptr objPtr;
+        };
+

         jsi::Object varObjectToJSIObject(juce::DynamicObject *v, jsi::Runtime &runtime)
         {
-            jassert(v);
-
-            //TODO: Should we be using jsi::HostObject here and binding to DynamicObject? Probably
-            //      in a full implementation.
-            //auto ho = std::make_shared<jsi::HostObject>();
-            jsi::Object jsiObj(runtime);
-
-            for (auto &prop : v->getProperties())
-            {
-                jsiObj.setProperty(runtime, prop.name.getCharPointer(), varToJSIValue(prop.value, runtime));
-            }
-
-            return jsiObj;
-            //return jsi::Object::createFromHostObject(runtime, ho);
+            auto obj = std::make_shared<JSIDynamicObject>(v);
+            return jsi::Object::createFromHostObject(runtime, obj);
         }
```

This approach doesn't work quite as nicely for the `jsiObjectToVarObject` side. I'm not sure whether guarding against lifetime issues is something we need to protect against here or if the current implementation is OK. What do you think? This is mainly important when users push some custom DynamicObject instance through to JS which invokes native methods etc. etc.

Reply:
Good question. So, in the js->native direction, most of the work that goes into marshaling an object to a juce::DynamicObject will copy primitive values (even if deeply nested). I think functions are really the only place to be careful, and our duktape approach gets by with this by sticking a global reference to the function in one of the private stashes so that the function won't disappear.

Here too we might be ok for functions, under one condition... Above, when we `std::move` the jsi::Function into the capture group of the juce::var::NativeFunction, that capture group will stay alive as long as somebody's holding a reference to the NativeFunction. So, if the `jsi::Function` itself implements the reference counting for the js func in the runtime, we won't drop that last count until the last reference to the NativeFunction in cpp drops. In that case I think we should be fine. Worth checking that detail of jsi::Function though.

Finally, for the native->js direction, yes you're super right, but I think this is not necessarily our responsibility. Let's make sure that wherever we push native->js for internal use, we are careful to protect lifetimes. But we don't need to jump through hoops to protect the end user from that case, I think the end user needs to be aware of their own lifetime management as they push values from native->js.

So all in all, I think the current approach is totally fine, and we should maybe just check that detail of jsi::Function

Reply:
Yep agreed. It was the end user case I was concerning myself over. i.e. the user pushes some DynamicObject instance through to JS which has functions bound to some C++ object. This would be an issue if the user hasn't subclassed `juce::DynamicObject` and kept the object alive for the duration if it's usage in JS, i.e. JS stores the object for usage later. 

I think I agree though. This is probably the user's responsibility as in our case we aren't storing these objects and basically work solely on the UI thread. So in the very rare cases where we push a DynamicObject instances with methods (I don't thnk we have any of these now) we're usually invoking the JS function immediately and the DynamicObject lifetime dies right after. 

I think leave implementation as is for now and we have this comment to refer to later should needs arise.